### ISSUE TITLE
Subrepos that have been removed are not converted correctly, cannot init submodules

### ISSUE THREAD

**Issue by jdanielp** [frej/fast-export#187]
Please see attached zip of an example mercurial repo with subrepos wherein one of the subrepos is not used in a certain branch after a certain change set.  

srt stands for "SubRepo Test".  The top level repo is srt_top, and it has two subrepos srt_suba and srt_subb.  Sorry for the naming; I apparently did not eat my Wheaties this morning. :)

For some time both subrepos are used by the top-level repo, but at change set e88d9415ea8e1a2a025b9df4c1054f2c2787cc72 srt_suba is removed from use by removing its entry in .hgsub.

I cloned out the subrepos to their own folders outsidee srt_top.  I converted the subrepos first, then the top level via the steps in README-Submodules.md

When I do the following in the git repo
`git checkout v3`
`git submodule update --init`

I receive the error:
> fatal: No url found for submodule path 'srt_suba' in .gitmodules

This also happens when using [Sourcetree](https://www.sourcetreeapp.com/) to navigate between branches.  This is the full output when attempting to checkout v3 branch:

> git -c diff.mnemonicprefix=false -c core.quotepath=false --no-optional-locks checkout v3 --progress
> M	srt_subb
> Switched to branch 'v3'
> git -c  diff.mnemonicprefix=false -c core.quotepath=false --no-optional-locks submodule update --init
> fatal: No url found for submodule path 'srt_suba' in .gitmodules
> Completed with errors, see above.

I have found some forum posts like [this](https://community.atlassian.com/t5/Sourcetree-questions/SourceTree-says-No-url-found-for-submodule-path/qaq-p/628869) and [this](https://stackoverflow.com/questions/4185365/no-submodule-mapping-found-in-gitmodule-for-a-path-thats-not-a-submodule) and doing the "git rm --cached X" will work, but only forward in time.  I tried re-writing history to add a commit that does the "git rm...." with "git rebase -i" via [this post](https://stackoverflow.com/questions/32315156/how-to-inject-a-commit-between-some-two-arbitrary-commits-in-the-past) but the interactive rebase had some merge problems in my actual repo and I'd rather not have to re-write history every changeset that removes a subrepo.

Please let me know if there is anything you need from me.  I am in charge of converting our mercurial repos to git repos because [bitbucket](https://bitbucket.org/blog/sunsetting-mercurial-support-in-bitbucket).  Right now I'm not familiar enough with the hg-fast-export tool to make the change myself and submit a pull request.  I am happy to beta test, though I suspect the attached example will be sufficient for verifying the first draft.

[srt_top.zip](https://github.com/frej/fast-export/files/4060496/srt_top.zip)
Edit: Adding 
[srt_top_submodule_mappings.txt](https://github.com/frej/fast-export/files/4073560/srt_top_submodule_mappings.txt)

**Comment by jdanielp** [comment]
Looks like I goofed when I created the example repo.  when running hg-fast-export.sh against the example repo I provide there is an error:
> fatal: Unsupported command: srt_suba = srt_suba
I have to figure out how to deal with subrepo paths in the .hgsub file.  We use [subpaths] in production.  I confirmed the error is not in my production repo, just in the sample I provide.  I will try to make a new example repo quickly.

**Comment by frej** [comment]
@MokhamedDakhraui do you have any ideas?

**Comment by MokhamedDakhraui** [comment]
Figured out what caused such behavior. If you look in mercurial commit 9, where subrepo `srt_suba` had been deleted, you can notice that file `.hgsubstate` still contains `srt_suba` record. The thing is that hg-fast-export converts submodules according that file.

Problem should be solved if you completely remove subrepo `srt_suba` from commit 9 (also from `.hgsubstate`).

Hope this'll help

**Comment by alexey-lcard** [comment]
But how can we change .hgsubstate in history commits of already existing repository without changing commits ids (this repository can oneself be subrepository in another top level repository)?
It seems strange that hg not remove automatically lines in .hgsubstate for subrepos removed from .hgsub, but hg works correctly with this problem and ignores this extra lines in .hgsubstate.

How can we convert existing working repositories with this problem to working git repositories?
Maybe hg-fast-export must first check .hgsub and use only lines in .hgsubstate that exists in .hgsub?

**Comment by jdanielp** [comment]
@MokhamedDakhraui  I agree with @alexey-lcard that modifying the history to remove the entry from `.hgsubstate` is not preferable.  hg works fine with this situation.  

I have been given the OK by my boss to work this issue, so I have time to spend on a fix.  I'm currently debugging through hg-fast-export.py to learn how things work.

One thing I noticed is that it is possible to determine what subrepos have been removed.  For example in hg-fasy-export.py line  528, when `rev` is 8 `ctx.substate` contains two entries corresponding to `srt_suba` and `srt_subb`.  At `rev` 9, `ctx.substate` only contains the entry for `srt_subb`.  I think we can use `ctx.substate` in `export_commit` to decide whether or not to add the submodule back in.

I'll keep investigating and provide an update here once I know more.
